// helpers/helpers.js
var hasDB = 0;

function initDB(done) {
	if (hasDB) done();
	else {
		hasDB=1;	
		request.post('/tools/db_init.php?step=upload&data=testdata.sql')
			.expect(200)
			.end(done);
	}
}

function inspector() {	
	var fxns = {
		done: function () {},
		body: function (res) {
			if (res && typeof res.body=='string') console.log(JSON.stringify(res.status)+'  '+JSON.stringify(res.body)); 
		},
	};
	
	function main(arg1, done) {
		if (arguments.length==2 && typeof arg2=='function') fxns.done = arg2;
		
		if (typeof arg1=='function') fxns.done = arg1;
		else if (typeof arg1=='string' && fxns[arg1]) return fxns[arg1];
		
		return fxns['body']; //default
	}
	
	return main;	
}

function compareKeys(a, required, optional) {
  var aKeys = Object.keys(a), ok = required.concat(optional); 
	
	var unexpected = aKeys.filter(function (x) { return ok.indexOf(x) < 0 });
	var missing = required.filter(function (x) { return aKeys.indexOf(x) < 0 });	
  if (unexpected.length || missing.length) return 'Unexpected properties: '+ JSON.stringify(unexpected) + ', mising: '+ JSON.stringify(missing)
}

function importTest(name, path) {
	describe(name, function () {
		require(path);
	});
}


exports.inspect = inspector();
exports.initDB = initDB;
exports.compareKeys = compareKeys;
exports.importTest = importTest;

exports.errHandler = (function errHandler(err) {
	console.log(err)
})

exports.wait = (function () {	
	var started=0, returned=0, mssg="", api, done, form;
	var indentedOk='    .....';
	
	function main(action) {
		started++;		
		replaceValuePatterns(action.inputs); 
		
		api.request(action).then(function (res) { 
			returned++;
			
			if (res.status != action.status) {
				mssg += action.label+": expecting="+ action.status +", got "+ res.status +".\n" + JSON.stringify(res.body['@graph'][0]);
			}
			else console.log(indentedOk + action.label +'-->')
			
			if (returned==started) done(mssg ? new Error(mssg) : null);	
		}, done)
	}
	
	function replaceValuePatterns(props) {
		for(var p in props) {
			if (typeof props[p]=='string') props[p] = props[p].replace('{xtime}', Date.now());
			if (props[p]=='{random}') props[p] = 1*props[p].replace('{random}', 40*Math.random().toFixed(1));
		}
	}
	
	main.reset = function (apiFxn, doneFxn) {
		started = 0; returned=0; mssg="";
		api = apiFxn;
		done = doneFxn;
	}
	
	main.orNot = function () {
		if (started==0) done();
	}	
	
	return main;
})()
